crates/containerd-shim-wasm/src/sandbox/containerd/client.rs

```diff
@@ -401,90 +382,122 @@ impl Client {
             return Ok((vec![], platform));
         };
 
-        log::info!("found manifest with WASM OCI image format.");
+        log::info!("found manifest with WASM OCI image format");
         // This label is unique across runtimes and version of the shim running
         // a precompiled component/module will not work across different runtimes or versions
         let (can_precompile, precompile_id) = match engine.can_precompile() {
             Some(precompile_id) => (true, precompile_label(T::name(), &precompile_id)),
             None => (false, "".to_string()),
         };
 
-        match image.labels.get(&precompile_id) {
-            Some(precompile_digest) if can_precompile => {
-                log::info!("found precompiled label: {} ", &precompile_id);
-                match self.read_content(precompile_digest) {
-                    Ok(precompiled) => {
-                        log::info!("found precompiled module in cache: {} ", &precompile_digest);
-                        return Ok((
-                            vec![WasmLayer {
-                                config: image_config_descriptor.clone(),
-                                layer: precompiled,
-                            }],
-                            platform,
-                        ));
-                    }
-                    Err(e) => {
-                        // log and continue
-                        log::warn!("failed to read precompiled module from cache: {}. Content may have been removed manually, will attempt to recompile", e);
-                    }
-                }
-            }
-            _ => {}
-        }
-
+        let image_info = self.get_info(&image_digest)?;
```

Comment:
Shouldn't this be checking content rather than the image?

One thing I ran into (and it also occurs with the "precompile each layer individually" API d8003bd) is that if i update a spin app to add another component, runwasi sees that the image needs precompilation again. I am assuming that is because the image digest changed so it doesn't see the label here.

This results in the Spin engine precompiling everything again, since runwasi passes the unprecompiled layers.

Reply:
I see what is happening -- and I don't think there is anything we can do about it. It looks like when you pull (`ctr image pull` ) the updated image, it removed the precompiled layers (even though it retains the labels on the original layers):

After pulling a Spin app with 2 components for the first time:
```
DIGEST									SIZE	AGE	LABELS
sha256:1a2671d57364ae2946fb4e2bae0c15e195644d39280a8bc976656af925935c6f	1.215kB	12 seconds	containerd.io/distribution.source.ghcr.io=kate-goldenring/spin-hello-and-kv-explorer
sha256:2314d31d6459abab359502dcd766bc50f09be84b8b46d4716441bf804f7a9bdf	664B	12 seconds	containerd.io/distribution.source.ghcr.io=kate-goldenring/spin-hello-and-kv-explorer,containerd.io/gc.ref.content.0=sha256:62410ce73ff98c3587580f7659d8dae08ec8f8e86b92f757b59bd52843d3f859,containerd.io/gc.ref.content.1=sha256:65bc286f8315746d1beecd2430e178f539fa487ebf6520099daae09a35dbce1d,containerd.io/gc.ref.content.2=sha256:1a2671d57364ae2946fb4e2bae0c15e195644d39280a8bc976656af925935c6f,containerd.io/gc.ref.content.config=sha256:aabedf0b094944b344cecbffcfa0a9172e905a5d156bdf67c4c6cbb6d27d1d92,containerd.io/gc.ref.content.precompile.0=sha256:f268328f6ef7f92da3950c2d1abc9b07fc240c021b9b3b354319a09c195c0eb0,containerd.io/gc.ref.content.precompile.1=sha256:dabc42e0d769727d23126b7908e0ea9dd4a93a7cd8867cb36615cbc990776ba8,runwasi.io/precompiled/spin/15627766957371465298=true
sha256:62410ce73ff98c3587580f7659d8dae08ec8f8e86b92f757b59bd52843d3f859	1.904MB	11 seconds	containerd.io/distribution.source.ghcr.io=kate-goldenring/spin-hello-and-kv-explorer,runwasi.io/precompiled/spin/15627766957371465298=sha256:f268328f6ef7f92da3950c2d1abc9b07fc240c021b9b3b354319a09c195c0eb0
sha256:65bc286f8315746d1beecd2430e178f539fa487ebf6520099daae09a35dbce1d	546.5kB	11 seconds	containerd.io/distribution.source.ghcr.io=kate-goldenring/spin-hello-and-kv-explorer,runwasi.io/precompiled/spin/15627766957371465298=sha256:dabc42e0d769727d23126b7908e0ea9dd4a93a7cd8867cb36615cbc990776ba8
sha256:aabedf0b094944b344cecbffcfa0a9172e905a5d156bdf67c4c6cbb6d27d1d92	78B	12 seconds	containerd.io/distribution.source.ghcr.io=kate-goldenring/spin-hello-and-kv-explorer
sha256:dabc42e0d769727d23126b7908e0ea9dd4a93a7cd8867cb36615cbc990776ba8	3.422MB	9 secondsrunwasi.io/precompiled/spin/15627766957371465298/original=sha256:65bc286f8315746d1beecd2430e178f539fa487ebf6520099daae09a35dbce1d
sha256:f268328f6ef7f92da3950c2d1abc9b07fc240c021b9b3b354319a09c195c0eb0	1.113MB	10 seconds	runwasi.io/precompiled/spin/15627766957371465298/original=sha256:62410ce73ff98c3587580f7659d8dae08ec8f8e86b92f757b59bd52843d3f859
```

After repulling the app that now has a third component:
```
DIGEST									SIZE	AGE		LABELS
sha256:290359e28f0878cac9793b37077c79337bdfe22e395cc809d4a99a40be2d8341	1.516kB	7 seconds	containerd.io/distribution.source.ghcr.io=kate-goldenring/spin-hello-and-kv-explorer
sha256:478f64c114b95dad0dbe88e9755384e30ba9c8bf62e76b93feceb85311607719	821B	8 seconds	containerd.io/distribution.source.ghcr.io=kate-goldenring/spin-hello-and-kv-explorer,containerd.io/gc.ref.content.0=sha256:62410ce73ff98c3587580f7659d8dae08ec8f8e86b92f757b59bd52843d3f859,containerd.io/gc.ref.content.1=sha256:65bc286f8315746d1beecd2430e178f539fa487ebf6520099daae09a35dbce1d,containerd.io/gc.ref.content.2=sha256:74e33464d5dae4246f077a821dd4c1514c26004e24f2b949ff4d1d95517f49e6,containerd.io/gc.ref.content.3=sha256:290359e28f0878cac9793b37077c79337bdfe22e395cc809d4a99a40be2d8341,containerd.io/gc.ref.content.config=sha256:aabedf0b094944b344cecbffcfa0a9172e905a5d156bdf67c4c6cbb6d27d1d92
sha256:62410ce73ff98c3587580f7659d8dae08ec8f8e86b92f757b59bd52843d3f859	1.904MB	About a minute	containerd.io/distribution.source.ghcr.io=kate-goldenring/spin-hello-and-kv-explorer,runwasi.io/precompiled/spin/15627766957371465298=sha256:f268328f6ef7f92da3950c2d1abc9b07fc240c021b9b3b354319a09c195c0eb0
sha256:65bc286f8315746d1beecd2430e178f539fa487ebf6520099daae09a35dbce1d	546.5kB	About a minute	containerd.io/distribution.source.ghcr.io=kate-goldenring/spin-hello-and-kv-explorer,runwasi.io/precompiled/spin/15627766957371465298=sha256:dabc42e0d769727d23126b7908e0ea9dd4a93a7cd8867cb36615cbc990776ba8
sha256:74e33464d5dae4246f077a821dd4c1514c26004e24f2b949ff4d1d95517f49e6	2.366MB	7 seconds	containerd.io/distribution.source.ghcr.io=kate-goldenring/spin-hello-and-kv-explorer
sha256:aabedf0b094944b344cecbffcfa0a9172e905a5d156bdf67c4c6cbb6d27d1d92	78B	About a minute	containerd.io/distribution.source.ghcr.io=kate-goldenring/spin-hello-and-kv-explorer
```

The precompiled layers (`dabc42e0d769727d23126b7908e0ea9dd4a93a7cd8867cb36615cbc990776ba8` and `f268328f6ef7f92da3950c2d1abc9b07fc240c021b9b3b354319a09c195c0eb0`) were removed and the new component layer `74e33464d5dae4246f077a821dd4c1514c26004e24f2b949ff4d1d95517f49e6` was added.

Reply:
> I can try adding the GC labels to the layers themselves as well. This might avoid the collection but since there is a new image the pre-compile will be invoked again.

@jsturtevant this would be a good addition as in the Spin shim we are checking for precompilation. This could also be done in a subsequent PR as an optimization if you are hesitant to iterate again

Reply:
I've added this extra GC label as well as a couple tests that cover the scenario

Reply:
That did the trick. Thank you!